packages/lodestar/src/db/api/beacon/repositories/blockArchive/abstract.ts

```diff
@@ -0,0 +1,116 @@
+    const blockRoot = this.signedBlockType.fields["message"].hashTreeRoot(value.message);
+    const slot = value.message.slot;
+    await Promise.all([
+      super.put(key, value),
+      storeRootIndex(this.db, slot, blockRoot),
+      storeParentRootIndex(this.db, slot, value.message.parentRoot),
+    ]);
+  }
+
+  public async batchPut(items: ArrayLike<IKeyValue<Slot, TBlock>>): Promise<void> {
+    await Promise.all([
+      super.batchPut(items),
+      Array.from(items).map((item) => {
+        const slot = item.value.message.slot;
+        const blockRoot = this.signedBlockType.fields["message"].hashTreeRoot(item.value.message);
+        return storeRootIndex(this.db, slot, blockRoot);
+      }),
+      Array.from(items).map((item) => {
+        const slot = item.value.message.slot;
+        const parentRoot = item.value.message.parentRoot;
+        return storeParentRootIndex(this.db, slot, parentRoot);
+      }),
+    ]);
+  }
+
+  public async batchPutBinary(items: ArrayLike<IKeyValueSummary<Slot, Buffer, IBlockSummary>>): Promise<void> {
+    await Promise.all([
+      super.batchPutBinary(items),
+      Array.from(items).map((item) => storeRootIndex(this.db, item.summary.slot, item.summary.blockRoot)),
+      Array.from(items).map((item) => storeParentRootIndex(this.db, item.summary.slot, item.summary.parentRoot)),
+    ]);
+  }
+
+  public async remove(value: TBlock): Promise<void> {
+    await Promise.all([
+      super.remove(value),
+      deleteRootIndex(this.db, this.signedBlockType.fields["message"] as ContainerType<TBlock["message"]>, value),
+      deleteParentRootIndex(this.db, value),
+    ]);
+  }
+
+  public async batchRemove(values: ArrayLike<TBlock>): Promise<void> {
+    await Promise.all([
+      super.batchRemove(values),
+      Array.from(values).map((value) =>
+        deleteRootIndex(this.db, this.signedBlockType.fields["message"] as ContainerType<TBlock["message"]>, value)
+      ),
+      Array.from(values).map((value) => deleteParentRootIndex(this.db, value)),
+    ]);
+  }
+
+  public decodeKey(data: Buffer): number {
+    return bytesToInt((super.decodeKey(data) as unknown) as Uint8Array, "be");
+  }
+
+  public getId(value: TBlock): Slot {
+    return value.message.slot;
+  }
+
+  public async values(opts?: IBlockFilterOptions): Promise<TBlock[]> {
+    return all(this.valuesStream(opts));
+  }
+
+  public async *valuesStream(opts?: IBlockFilterOptions): AsyncIterable<TBlock> {
+    const dbFilterOpts = this.dbFilterOptions(opts);
+    const firstSlot = dbFilterOpts.gt
+      ? this.decodeKey(dbFilterOpts.gt) + 1
+      : dbFilterOpts.gte
+      ? this.decodeKey(dbFilterOpts.gte)
+      : null;
+    if (firstSlot === null) throw Error("specify opts.gt or opts.gte");
+    const valuesStream = super.valuesStream(opts);
+    const step = (opts && opts.step) || 1;
+
+    for await (const value of valuesStream) {
+      if ((value.message.slot - firstSlot) % step === 0) {
+        yield value;
+      }
+    }
```

Comment:
Not completely related to this PR but if someone sends a blocksByRangeRequest to us with
```
startSlot: 0,
count: 2,
step: 360000
```
we would read the entire block DB?

Reply:
probably. We can add some validation to blocksByRange to limit step size.

Reply:
Or if step size is > stepSizeLimit don't stream and fetch one by one